main_windows.go

```diff
@@ -86,3 +90,43 @@ func obtainLock() {
 
 	deferredFuncs = append(deferredFuncs, func() { unlockFileEx(f.Fd(), 1, 0, &syscall.Overlapped{}); f.Close(); os.Remove(lockFile) })
 }
+
+type service struct {
+	ctx context.Context
+	run func(context.Context)
+}
+
+func (s *service) Execute(args []string, r <-chan svc.ChangeRequest, status chan<- svc.Status) (svcSpecificEC bool, exitCode uint32) {
+	status <- svc.Status{State: svc.StartPending}
+	ctx, cncl := context.WithCancel(s.ctx)
+	defer cncl()
+	done := make(chan struct{})
+
+	go func() {
+		s.run(ctx)
+		close(done)
+	}()
+	status <- svc.Status{State: svc.Running, Accepts: svc.AcceptStop | svc.AcceptShutdown}
+
+	for {
+		select {
+		case <-done:
+			status <- svc.Status{State: svc.StopPending}
+			return
+		case c := <-r:
+			switch c.Cmd {
+			case svc.Interrogate:
+				status <- c.CurrentStatus
+			case svc.Stop, svc.Shutdown:
+				cncl()
+			default:
+			}
+		}
+	}
+}
+
+func runService(ctx context.Context, run func(context.Context)) {
```

Comment:
same here. we own this and are not creating a general-purpose library. don't take functions as arguments.